src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -581,20 +581,23 @@ trait Namers extends MethodSynthesis {
         }
       }
 
-      def noDuplicates(names: List[Name], check: DuplicatesErrorKinds.Value) {
-        def loop(xs: List[Name]): Unit = xs match {
+      def noDuplicatesBy(namer: ImportSelector => Name, check: DuplicatesErrorKinds.Value): Unit = {
+        def canSkip(name: Name, sels: List[ImportSelector]): Boolean =
+          name == null || name == nme.WILDCARD || sels.forall(name ne namer(_))
+        @tailrec
+        def loop(xs: List[ImportSelector]): Unit = xs match {
           case Nil      => ()
           case hd :: tl =>
-            if (hd == nme.WILDCARD || !(tl contains hd)) loop(tl)
-            else DuplicatesError(tree, hd, check)
+            if (canSkip(namer(hd), tl)) loop(tl)
+            else DuplicatesError(tree, namer(hd), check)
         }
-        loop(names filterNot (x => x == null || x == nme.WILDCARD))
+        loop(selectors)
       }
       selectors foreach checkSelector
 
       // checks on the whole set
-      noDuplicates(selectors map (_.name), RenamedTwice)
-      noDuplicates(selectors map (_.rename), AppearsTwice)
+      noDuplicatesBy(  nameOfImportSelector, RenamedTwice)
+      noDuplicatesBy(renameOfImportSelector, AppearsTwice)
```

Comment:
This is a lot less clear than `_.name`/`_.rename`, and incurs a field access... what's the benefit here?

Reply:
If your question is about why I am using this constant lambda literals `nameOfImportSelector` and `renameOfImportSelector`, I am following a practice that I have seen in other places of the compiler, of extracting `Function1` objects outside of the main loop, so as to avoid allocating the `Function1` object more than once per compiler run.

Reply:
That practice is often misguided these days; if a function expression doesn't close over anything, the bootstrap method which makes the instance will cache it at the call site; this will cause us to have to read it from the outer class.